src/libsystemd-network/sd-dhcp-server.c

```diff
@@ -1306,3 +1369,21 @@ int sd_dhcp_server_set_relay_target(sd_dhcp_server *server, const struct in_addr
         server->relay_target = *address;
         return 1;
 }
+
+int sd_dhcp_server_set_relay_agent_information(
+                sd_dhcp_server *server,
+                const char *agent_circuit_id,
+                const char *agent_remote_id) {
+        int r;
+
+        assert_return(server, -EINVAL);
+
+        if (relay_agent_information_length(agent_circuit_id, agent_remote_id) > 255)
+                return -ENOBUFS;
+
+        r = free_and_strdup(&server->agent_circuit_id, agent_circuit_id);
+        if (r < 0)
+                return r;
+
+        return free_and_strdup(&server->agent_remote_id, agent_remote_id);
```

Comment:
When the second `free_and_strdup()` fails, the information in sd_dhcp_server is partially updated. Please do something like the following:
```
_cleanup_free_ char *circuit = NULL, *remote = NULL;
int r;

if (... > 255)
        return -EINVAL;

circuit = strdup(agent_circuit_id);
if (!circuit)
        return -ENOMEM;

remote = strdup(agent_remote_id);
if (!remote)
        return -ENOMEM; 

free_and_replace(server->agent_circuit_id, circuit);
free_and_replace(server->agent_remote_id, remote);
return 0;
```